// firebase-tools/lib/deploy/functions/services/ailogic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AILogicService = exports.AI_LOGIC_EVENTS = exports.AI_LOGIC_AFTER_GENERATE_CONTENT = exports.AI_LOGIC_BEFORE_GENERATE_CONTENT = void 0;
exports.isAILogicEvent = isAILogicEvent;
const backend = require("../backend");
const error_1 = require("../../../error");
const ailogicApi = require("../../../gcp/ailogic");
exports.AI_LOGIC_BEFORE_GENERATE_CONTENT = "firebase.vertexai.v1beta.beforeGenerateContent";
exports.AI_LOGIC_AFTER_GENERATE_CONTENT = "firebase.vertexai.v1beta.afterGenerateContent";
exports.AI_LOGIC_EVENTS = [
    exports.AI_LOGIC_BEFORE_GENERATE_CONTENT,
    exports.AI_LOGIC_AFTER_GENERATE_CONTENT,
];
function isAILogicEvent(endpoint) {
    if (!backend.isBlockingTriggered(endpoint)) {
        return false;
    }
    return exports.AI_LOGIC_EVENTS.includes(endpoint.blockingTrigger.eventType);
}
class AILogicService {
    constructor() {
        this.ensureTriggerRegion = () => Promise.resolve();
        this.name = "ailogic";
        this.api = "firebasevertexai.googleapis.com";
    }
    validateTrigger(endpoint, wantBackend) {
        if (!isAILogicEvent(endpoint)) {
            return;
        }
        const eventType = endpoint.blockingTrigger.eventType;
        const regionalWebhook = !!endpoint.blockingTrigger.options?.regionalWebhook;
        const conflict = backend.allEndpoints(wantBackend).some((ep) => {
            if (!isAILogicEvent(ep)) {
                return false;
            }
            if (ep.blockingTrigger.eventType !== eventType || ep.id === endpoint.id) {
                return false;
            }
            if (regionalWebhook) {
                return ep.blockingTrigger.options?.regionalWebhook && ep.region === endpoint.region;
            }
            else {
                return !ep.blockingTrigger.options?.regionalWebhook;
            }
        });
        if (conflict) {
            if (regionalWebhook) {
                throw new error_1.FirebaseError(`Can only create at most one regional AI Logic Trigger for ${eventType} in region ${endpoint.region}`);
            }
            else {
                throw new error_1.FirebaseError(`Can only create at most one global AI Logic Trigger for ${eventType}`);
            }
        }
    }
    async registerTrigger(ep) {
        if (!isAILogicEvent(ep)) {
            return;
        }
        await ailogicApi.upsertBlockingFunction(ep);
    }
    async unregisterTrigger(ep) {
        if (!isAILogicEvent(ep)) {
            return;
        }
        await ailogicApi.deleteBlockingFunction(ep);
    }
}
exports.AILogicService = AILogicService;
